jolt-core/src/dag/state_manager.rs

```diff
@@ -0,0 +1,195 @@
+pub const BIG_ENDIAN: Endianness = false;
+pub const LITTLE_ENDIAN: Endianness = true;
+
+#[derive(Clone, Debug)]
+pub struct OpeningPoint<const E: Endianness, F: JoltField> {
+    pub r: Vec<F>,
+}
+
+impl<const E: Endianness, F: JoltField> OpeningPoint<E, F> {
+    pub fn new(r: Vec<F>) -> Self {
+        Self { r }
+    }
+
+    pub fn endianness(&self) -> &'static str {
+        if E == BIG_ENDIAN {
+            "big"
+        } else {
+            "little"
+        }
+    }
+
+    pub fn match_endianness<const SWAPPED_E: Endianness>(&self) -> OpeningPoint<SWAPPED_E, F>
+    where
+        F: Clone,
+    {
+        let mut reversed = self.r.clone();
+        if E != SWAPPED_E {
+            reversed.reverse();
+        }
+        OpeningPoint::<SWAPPED_E, F>::new(reversed)
+    }
+}
+
+impl<F: JoltField> From<Vec<F>> for OpeningPoint<LITTLE_ENDIAN, F> {
+    fn from(r: Vec<F>) -> Self {
+        Self::new(r)
+    }
+}
+
+impl<F: JoltField> From<Vec<F>> for OpeningPoint<BIG_ENDIAN, F> {
+    fn from(r: Vec<F>) -> Self {
+        Self::new(r)
+    }
+}
+
+impl<F: JoltField> Index<JoltR1CSInputs> for Openings<F> {
+    type Output = Option<F>;
+
+    fn index(&self, index: JoltR1CSInputs) -> &Self::Output {
+        &self[&OpeningsKeys::SpartanZ(index)].1
+    }
+}
+
+#[derive(Hash, PartialEq, Eq, Copy, Clone, Debug)]
+pub enum OpeningsKeys {
+    SpartanZ(JoltR1CSInputs),
+    InstructionTypeFlag(usize),
+    InstructionRa(usize),
+    OuterSumcheckClaims, // (Az, Bz, Cz)
+    OuterSumcheckR,      // r_cycle from outer sumcheck
+    OuterSumcheckRxVar,  // rx_var from outer sumcheck
+}
+
+pub type Openings<F> = HashMap<OpeningsKeys, (OpeningPoint<LITTLE_ENDIAN, F>, Option<F>)>;
+
+#[derive(Hash, PartialEq, Eq, Clone, Debug)]
+pub enum ProofKeys {
+    SpartanOuterSumcheck,
+    SpartanInnerSumcheck,
+    SpartanShiftSumcheck,
+    RegistersReadWrite(usize),
+    RamReadWrite(usize),
+    InstructionLookups(usize),
+}
+
+pub enum ProofData<F: JoltField, ProofTranscript: Transcript> {
+    Spartan(UniformSpartanProof<F, ProofTranscript>),
+    SpartanSumcheck(SumcheckInstanceProof<F, ProofTranscript>, Vec<F>, [F; 3]),
+}
+
+pub type Proofs<F, ProofTranscript> = HashMap<ProofKeys, ProofData<F, ProofTranscript>>;
+
+pub struct SpartanState<'a, F: JoltField> {
+    pub spartan_key: Option<&'a UniformSpartanKey<F>>,
+    pub constraint_builder: Option<&'a CombinedUniformBuilder<F>>,
+    pub input_polys: Option<Vec<MultilinearPolynomial<F>>>,
+}
+
+pub struct StateManager<
+    'a,
+    F: JoltField,
+    ProofTranscript: Transcript,
+    PCS: CommitmentScheme<ProofTranscript, Field = F>,
+> {
+    pub openings: Arc<Mutex<Openings<F>>>,
+    pub prover_accumulator: &'a mut ProverOpeningAccumulator<F, PCS, ProofTranscript>,
+    pub verifier_accumulator: &'a mut VerifierOpeningAccumulator<F, PCS, ProofTranscript>,
+    pub prover_transcript: RefCell<&'a mut ProofTranscript>,
+    pub verifier_transcript: RefCell<&'a mut ProofTranscript>,
```

Comment:
why do we need verifier_transcript if its the same as prover_transcript?

Reply:
I think just a single transcript here would suffice